py/server/deephaven_internal/auto_completer/_completer.py

```diff
@@ -253,3 +258,36 @@ def do_hover(
             hoverstring += '\n---\n' + wrap_plaintext(raw_docstring)
 
         return hoverstring.strip()
+
+    def check_recursion_limit(self):
+        """
+        Tests for python+jedi+numpy versions that are susceptible to a RecursionError/segfault issue, and lowers
+        the recursion limit, warning if the limit is raised externally
+        """
+        if not sys.version.startswith('3.9.') and not sys.version.startswith('3.10.'):
+            return
+
+        if not self.__can_jedi:
+            return
+
+        # numpy is a required dependency, tested for jedi above
+        if not version('numpy').startswith('2.0') or not version('jedi') == '0.19.1':
+            return
+
+        from . import MAX_RECURSION_LIMIT
+
+        if sys.getrecursionlimit() <= MAX_RECURSION_LIMIT:
+            return
+
+        sys.setrecursionlimit(MAX_RECURSION_LIMIT)
+
+        if not self.recursion_limit_already_warned:
+            self.recursion_limit_already_warned = True
+            warnings.warn(f"""Recursion limit has been set to {MAX_RECURSION_LIMIT} to avoid a known segfault in Python related to RecursionErrors.
+This limit will be set to {MAX_RECURSION_LIMIT} whenever autocomplete takes place to avoid this. Other steps you can
+take to avoid this:
+ * Use numpy 1.x
+ * Use Python 3.8 or 3.11+
+ * When available, use a newer version of Jedi
+ * Disable autocomplete
+See https://github.com/deephaven/deephaven-core/issues/5878 for more information.""")
```

Comment:
I think the fix definitely works, however the following approach might be cleaner to implement and a bit more robust, though not as flexible:
1. for py 3.9 and 3.10, restore the sys default recursion limit (2000)
2. in `do_completion`, add try/except to catch `exception RecursionError` and provide a meaningful message  as the above
3. we can go further to catch any other possible exception and provide some feedback to the user

Reply:
Unfortunately for step 2, we cannot catch the RecursionError - instead of emitting a RecursionError, libpython simply segfaults.

Your step 1 has the advantage of not emitting a warning _unless_ the user changes the recursion limit after autocomplete has run - but if they set the recursion limit _before_ we import jedi, then we would stomp on that change. Perhaps I also do a check before `import jedi` then to see if they had already set the limit higher?

There is also the gotcha that in order to expose `MAX_RECURSION_LIMIT` in the init module, I can't read it until _after_ the Completer has been instantiated (otherwise we have a cyclical reference between files). Are we okay with this not being user configurable?

I'll put up an example of this, and see what you think.